flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SchedulerImpl.java

```diff
@@ -562,4 +575,110 @@ private void releaseSharedSlot(
 	public boolean requiresPreviousExecutionGraphAllocations() {
 		return slotSelectionStrategy instanceof PreviousAllocationSlotSelectionStrategy;
 	}
+
+	@Override
+	public CompletableFuture<Collection<PhysicalSlotRequest.Result>> allocatePhysicalSlots(
+			final Collection<PhysicalSlotRequest> physicalSlotRequests,
+			final Time timeout) {
+
+		final PhysicalSlotRequestBulk slotRequestBulk = new PhysicalSlotRequestBulk(physicalSlotRequests);
+
+		final List<CompletableFuture<PhysicalSlotRequest.Result>> resultFutures = new ArrayList<>(physicalSlotRequests.size());
+		for (PhysicalSlotRequest request : physicalSlotRequests) {
+			final CompletableFuture<PhysicalSlotRequest.Result> resultFuture =
+				allocatePhysicalSlot(request, timeout).thenApply(result -> {
+					slotRequestBulk.markRequestFulfilled(
+						result.getSlotRequestId(),
+						result.getPhysicalSlot().getAllocationId());
+
+					return result;
+				});
+			resultFutures.add(resultFuture);
+		}
+
+		slotRequestBulkTracker.track(slotRequestBulk);
+		schedulePendingRequestBulkTimeoutCheck(slotRequestBulk, timeout);
+
+		return FutureUtils.combineAll(resultFutures)
+			.whenComplete((ignore, throwable) -> slotRequestBulkTracker.untrack(slotRequestBulk));
+	}
+
+	private CompletableFuture<PhysicalSlotRequest.Result> allocatePhysicalSlot(
+			final PhysicalSlotRequest physicalSlotRequest,
+			final Time timeout) {
+
+		final SlotRequestId slotRequestId = physicalSlotRequest.getSlotRequestId();
+		final SlotProfile slotProfile = physicalSlotRequest.getSlotProfile();
+
+		final Optional<SlotAndLocality> availablePhysicalSlot = tryAllocateFromAvailable(slotRequestId, slotProfile);
```

Comment:
I think that the majority part of `SchedulerImpl` will be needed anymore in the future.
Introducing a separate interface like `BulkSlotProvider ` might make it easier for us to drop the deprecated components in the future.

> Do you think we can reuse SchedulerImpl in future?
I think not.

> Would just duplicating tryAllocateFromAvailable/cancelSlotRequest in BulkSlotProviderImpl bring less confusion in future?
I think yes. Let's do it this way.

> Will we need single slot provider eventually for pipeline region scheduling at all?
I think not.

Reply:
done.

Reply:
I just realized that `BulkSlotProvider ` is actually making things more complicated because we will need to change how we create it and its factory in `JobMaster` and `DefaultJobManagerRunnerFactory`.
`BulkSlotProvider` needs to know main thread executor and slot pool, which makes it hard to create it in SchedulerNG internally.
Then we will need to keep `BulkSlotProvider` and `Scheduler`/`SlotProvider` at the same time in JobMaster before we can completely drop `SchedulerImpl`.

I think it would be easier to just keep the bulk slot allocation logics `SchedulerImpl`. We can try to make them less coupled with existing logics that will be dropped in the future.
WDYT? @azagrebin

Reply:
I find one way that we can still have a separate `BulkSlotProvider` and do not need to change `Scheduler`/`SlotProvider ` creating and initializing codes.
That is, we still enables `SlotProvider` for bulk slot allocation. However, `SchedulerImpl` creates a `BulkSlotProviderImpl` internally and dedicates bulk slot allocation requests to it.
See 395c6cba8c463cffe2cf3cc70d91e6c6ff0dbb9f